### PR TITLE
Add support for nonNamespaced generated reconcilers

### PR DESCRIPTION
**Pull request by nachocano**
- As with genClient, added `genreconciler:nonNamespaced` to provide cluster-scope resources generated reconcilers.

/assign @n3wscott

### CODE REVIEW

**Comment by mattmoor** [comment]
Superficially this LGTM.  @n3wscott any thoughts since you've done the most here?

File: codegen/cmd/injection-gen/generators/packages.go [hunk 1]
```diff
@@ -169,9 +169,11 @@ func MustParseClientGenTags(lines []string) Tags {
 	_, ret.GenerateDuck = values["genduck"]
 
 	_, genRec := values["genreconciler"]
+	_, genRecNonNamespaced := values["genreconciler:nonNamespaced"]
```
Comment (n3wscott) [inline codegen/cmd/injection-gen/generators/packages.go]: Use the real client key, no need to add a new key. 
```
// +genclient:nonNamespaced
```

**Comment by n3wscott** [comment]
Rather than duplicating all the places we make the `Get` and `Update` calls, perhaps it would be better to do something like 

```
{{if .nonNamespaced}}
client := r.Client.{{.group}}{{.version}}().{{.type|apiGroup}}()
{{else}}
client := r.Client.{{.group}}{{.version}}().{{.type|apiGroup}}(existing.Namespace)
{{end}}

_, err = client.UpdateStatus(existing)
```

What do you think?


Also you will need to add a new resource that is generated at cluster scope to the test client to confirm this all compiles.

**Comment by nachocano** [comment]
@n3wscott all comments make sense. Thanks for reviewing, will try to get back to this later today and will let you know when is ready for another look

**Comment by n3wscott** [comment]
closed via https://github.com/knative/pkg/pull/1175

**Pull request closed without merging**